File: mine/circlequeue/queue.py
import logging

class QError(BaseException):
    def __init__(self, value, message):
        self.value = value
        self.message = message

    def __str__(self):
        # return repr(self.message)
        return self.message

class Queue:

    def __init__(self, capacity):
        self.queue = []
        self.front = 0
        self.back = 0
        self.capacity = capacity + 1
        for i in range(capacity + 1):
            self.queue.append(0)

    def Front(self):
        return self.queue[self.front + 1]

    def Back(self):
        return self.queue[self.back]

    def Push(self, element):
        if self.Full():
            raise QError(-1, 'Queue full')
        else:
            self.back = (self.back + 1) % self.capacity
            self.queue[self.back] = element

    def Pop(self):
        if self.Empty():
            raise QError(-2, 'Queue empty')
        else:
            self.front = (self.front + 1) % self.capacity

    def Full(self):
        return (self.back + 1) % self.capacity == self.front

    def Empty(self):
        return self.front == self.back


if __name__ == '__main__':
    q = Queue(5)
    print(q.Empty())
    print(q.Full())

    try:
        q.Push(3)
        print(q.Empty())
        print(q.Full())
        q.Push(4)
        q.Push(5)
        q.Push(6)
        q.Push(7)
        print(q.Empty())
        print(q.Full())

        q.Pop()
        print(q.Empty())
        print(q.Full())
        q.Pop()
        q.Pop()
        q.Pop()
        print(q.Empty())
        print(q.Full())
        q.Pop()
        print(q.Empty())
        print(q.Full())

        #queue already empty
        q.Pop()
    except QError as e:
        # 通过配置，logging还可以把错误记录到日志文件里，方便事后排查
        logging.exception(e)

    print('END')

File: mine/circlequeue/queue.rs
#![allow(dead_code)]
use std::fmt::Display;

fn main() {
    let mut q: Queue<i32> = Queue::new(4);
    q.build(&0);
    println!("{}", q.empty());
    println!("{}", q.full());
    q.push(&3);
    println!("{}", q.empty());
    println!("{}", q.full());
    q.push(&4);
    q.push(&5);
    q.push(&6);
    println!("{}", q.empty());
    println!("{}", q.full());
    q.dump();
    q.pop();
    println!("-----------------");
    q.dump();
    q.push(&7);
    q.dump();
    println!("{}", q.empty());
    println!("{}", q.full());
    q.pop();
    q.pop();
    q.pop();
    q.pop();
    println!("{}", q.empty());
    println!("{}", q.full());
}

struct Queue<T> {
    queue: Vec<T>,
    front: usize,
    back: usize,
    capacity: usize
}

impl<T> Queue<T> where T: Copy + Display {
    fn new(capacity: usize) -> Queue<T> {
        Queue {
            queue: Vec::with_capacity(capacity + 1),
            front: 0,
            back: 0,
            capacity: capacity + 1
        }
    }

    fn build(&mut self, t: &T) {
        for _i in 0..self.capacity {
            self.queue.push(*t);
        }
    }

    fn full(&self) -> bool {
        (self.back + 1) % self.capacity == self.front
    }

    fn empty(&self) -> bool {
        self.front == self.back
    }

    fn front(&self) -> &T {
        &self.queue[self.front + 1]
    }

    fn back(&self) -> &T {
        &self.queue[self.back]
    }

    fn push(&mut self, t: &T) -> bool {
        if self.full() {
            return true;
        }
        self.back = (self.back + 1) % self.capacity;
        self.queue[self.back] = *t;
        true
    }

    fn pop(&mut self) -> bool {
        if self.empty() {
            return false;
        }
        self.front = (self.front + 1) % self.capacity;
        true
    }

    fn dump(&self) {
        if self.front < self.back {
            for i in self.front + 1..self.back + 1 {
                println!("{}", self.queue[i]);
            }
        } else if self.front > self.back {
            println!("~~~~~~~~~~~~~~{}, {}", self.front + 1, self.queue.capacity());
            for i in self.front + 1..self.queue.capacity() {
                println!("{}", self.queue[i]);
            }

            println!("~~~~~~~~~~~~~~{}, {}", self.front + 1, self.queue.capacity());

            for i in 0..self.back + 1 {
                println!("{}", self.queue[i]);
            }
        }
    }
}
